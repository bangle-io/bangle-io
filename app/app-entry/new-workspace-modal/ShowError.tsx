import React from 'react';

import { useActionContext } from '@bangle.io/action-context';
import { CORE_PALETTES_TOGGLE_WORKSPACE_PALETTE } from '@bangle.io/constants';
import { safeRequestAnimationFrame } from '@bangle.io/utils';

import {
  ERROR_PICKING_DIRECTORY_ERROR,
  INVALID_WORKSPACE_NAME_ERROR,
  UNKNOWN_ERROR,
  WORKSPACE_AUTH_REJECTED_ERROR,
  WORKSPACE_NAME_ALREADY_EXISTS_ERROR,
  WorkspaceCreateErrorTypes,
} from './common';

export function ShowError({
  errorType,
  closeModal,
}: {
  errorType: WorkspaceCreateErrorTypes | undefined;
  closeModal: () => void;
}) {
  const { dispatchAction } = useActionContext();
  if (!errorType) {
    return null;
  }

  let content;

  switch (errorType) {
    case WORKSPACE_NAME_ALREADY_EXISTS_ERROR: {
      content = (
        <span>
          A workspace with the same name already exists.{' '}
          <button
            className="underline"
            onClick={() => {
              closeModal();
              safeRequestAnimationFrame(() => {
                dispatchAction({
                  name: CORE_PALETTES_TOGGLE_WORKSPACE_PALETTE,
                });
              });
            }}
          >
            Click here
          </button>{' '}
          to open it
        </span>
      );
      break;
    }

    case ERROR_PICKING_DIRECTORY_ERROR: {
      content = (
        <span>
          There was an error opening your notes folder. Please make sure your
          notes folder inside a common location like Documents or Desktop.
        </span>
      );
      break;
    }
    case INVALID_WORKSPACE_NAME_ERROR: {
      content = (
        <span>
          Invalid workspace name. Workspace cannot have <code>:</code> in its
          name.
        </span>
      );
      break;
    }
    case WORKSPACE_AUTH_REJECTED_ERROR: {
      content = (
        <span>
          Bangle.io was denied access to your notes. Please try again and press
          allow Bangle.io access to your locally saved notes.
        </span>
      );
      break;
    }
    case UNKNOWN_ERROR: {
      content = (
        <span>
          An unknown error occurred. Please reload window and try again. If the
          problem still persists open an issue at{' '}
          <a
            href="https://github.com/bangle-io/bangle-io-issues"
            className="underline"
          >
            Github
          </a>{' '}
        </span>
      );
      break;
    }

    default: {
      // hack to catch switch slipping
      let val: never = errorType;
      throw new Error('Unknown error type ' + val);
    }
  }

  return (
    <div
      className="w-full px-4 text-center"
      data-testid={errorType}
      style={{
        backgroundColor: 'var(--severity-error-color)',
        color: 'white',
      }}
    >
      {content}
    </div>
  );
}
