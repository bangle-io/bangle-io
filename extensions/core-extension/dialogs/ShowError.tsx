import React from 'react';

import { nsmApi2 } from '@bangle.io/api';
import { CorePalette } from '@bangle.io/constants';
import { vars } from '@bangle.io/css-vars';
import { safeRequestAnimationFrame } from '@bangle.io/utils';

import type { WorkspaceCreateErrorTypes } from './common';
import {
  CLICKED_TOO_SOON_ERROR,
  ERROR_PICKING_DIRECTORY_ERROR,
  INVALID_WORKSPACE_NAME_ERROR,
  UNKNOWN_ERROR,
  WORKSPACE_AUTH_REJECTED_ERROR,
  WORKSPACE_NAME_ALREADY_EXISTS_ERROR,
} from './common';

export function ShowError({
  errorType,
  closeModal,
}: {
  errorType: WorkspaceCreateErrorTypes | undefined;
  closeModal: () => void;
}) {
  if (!errorType) {
    return null;
  }

  let content, title;

  switch (errorType) {
    case WORKSPACE_NAME_ALREADY_EXISTS_ERROR: {
      title = 'A workspace with the same name already exists.';
      content = (
        <div>
          <button
            className="underline"
            onClick={() => {
              closeModal();
              safeRequestAnimationFrame(() => {
                nsmApi2.ui.togglePalette(CorePalette.Workspace);
              });
            }}
          >
            Click here
          </button>{' '}
          to open it
        </div>
      );
      break;
    }

    case ERROR_PICKING_DIRECTORY_ERROR: {
      title = 'There was an error opening your notes folder.';
      content = (
        <div>
          Please make sure your notes folder inside a common location like
          Documents or Desktop.
        </div>
      );
      break;
    }

    case CLICKED_TOO_SOON_ERROR: {
      title = 'That didnâ€™t work';
      content = <div>Please try clicking the Browse button again.</div>;
      break;
    }
    case INVALID_WORKSPACE_NAME_ERROR: {
      title = 'Invalid workspace name';
      content = (
        <div>
          Workspace cannot have <code>:</code> in its name.
        </div>
      );
      break;
    }
    case WORKSPACE_AUTH_REJECTED_ERROR: {
      title = 'Bangle.io was denied access to your notes.';
      content = (
        <div>
          Please try again and press <i>allow</i> to let Bangle.io access your
          locally saved notes.
        </div>
      );
      break;
    }
    case UNKNOWN_ERROR: {
      title = 'An unknown error occurred.';
      content = (
        <div>
          Please reload window and try again. If the problem still persists open
          an issue at{' '}
          <a
            href="https://github.com/bangle-io/bangle-io/issues"
            className="underline"
          >
            Github
          </a>{' '}
        </div>
      );
      break;
    }

    default: {
      // hack to catch switch slipping
      let val: never = errorType;
      throw new Error('Unknown error type ' + val);
    }
  }

  return (
    <div
      className="w-full m-1 px-5 py-3 text-center rounded"
      data-testid={errorType}
      style={{
        backgroundColor: vars.color.critical.solidFaint,
        color: vars.color.critical.solidText,
      }}
    >
      <div className="font-semibold text-left">{title}</div>
      <div className="text-left">{content}</div>
    </div>
  );
}
