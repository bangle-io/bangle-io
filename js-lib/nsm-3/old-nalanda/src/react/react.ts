import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
