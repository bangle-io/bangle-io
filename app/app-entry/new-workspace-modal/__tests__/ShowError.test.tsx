/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react';
import React from 'react';

import {
  ERROR_PICKING_DIRECTORY_ERROR,
  INVALID_WORKSPACE_NAME_ERROR,
  UNKNOWN_ERROR,
  WORKSPACE_AUTH_REJECTED_ERROR,
  WORKSPACE_NAME_ALREADY_EXISTS_ERROR,
} from '../common';
import { ShowError } from '../ShowError';

jest.mock('@bangle.io/slice-ui', () => {
  const otherThings = jest.requireActual('@bangle.io/slice-ui');

  return {
    ...otherThings,
    useUIManagerContext: jest.fn(() => ({})),
  };
});

test('renders correctly UNKNOWN_ERROR', () => {
  let result = render(
    <div>
      <ShowError errorType={UNKNOWN_ERROR} closeModal={jest.fn()}></ShowError>
    </div>,
  );

  expect(result.container.innerHTML).toContain('unknown error');
  expect(result.container).toMatchSnapshot();
});

test('renders correctly ERROR_PICKING_DIRECTORY_ERROR', () => {
  let result = render(
    <div>
      <ShowError
        errorType={ERROR_PICKING_DIRECTORY_ERROR}
        closeModal={jest.fn()}
      ></ShowError>
    </div>,
  );

  expect(result.container).toMatchSnapshot();
});

test('renders correctly INVALID_WORKSPACE_NAME_ERROR', () => {
  let result = render(
    <div>
      <ShowError
        errorType={INVALID_WORKSPACE_NAME_ERROR}
        closeModal={jest.fn()}
      ></ShowError>
    </div>,
  );

  expect(result.container).toMatchSnapshot();
});

test('renders correctly WORKSPACE_AUTH_REJECTED_ERROR', () => {
  let result = render(
    <div>
      <ShowError
        errorType={WORKSPACE_AUTH_REJECTED_ERROR}
        closeModal={jest.fn()}
      ></ShowError>
    </div>,
  );

  expect(result.container).toMatchSnapshot();
});

test('renders correctly WORKSPACE_NAME_ALREADY_EXISTS_ERROR', () => {
  let result = render(
    <div>
      <ShowError
        errorType={WORKSPACE_NAME_ALREADY_EXISTS_ERROR}
        closeModal={jest.fn()}
      ></ShowError>
    </div>,
  );

  expect(result.container).toMatchSnapshot();
});
