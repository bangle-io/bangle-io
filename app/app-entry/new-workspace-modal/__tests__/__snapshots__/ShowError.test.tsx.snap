// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly ERROR_PICKING_DIRECTORY_ERROR 1`] = `
<div>
  <div>
    <div
      class="w-full px-4 text-center"
      data-testid="ERROR_PICKING_DIRECTORY"
      style="color: white;"
    >
      <span>
        There was an error opening your notes folder. Please make sure your notes folder inside a common location like Documents or Desktop.
      </span>
    </div>
  </div>
</div>
`;

exports[`renders correctly INVALID_WORKSPACE_NAME_ERROR 1`] = `
<div>
  <div>
    <div
      class="w-full px-4 text-center"
      data-testid="INVALID_WORKSPACE_NAME"
      style="color: white;"
    >
      <span>
        Invalid workspace name. Workspace cannot have 
        <code>
          :
        </code>
         in its name.
      </span>
    </div>
  </div>
</div>
`;

exports[`renders correctly UNKNOWN_ERROR 1`] = `
<div>
  <div>
    <div
      class="w-full px-4 text-center"
      data-testid="UNKNOWN_ERROR"
      style="color: white;"
    >
      <span>
        An unknown error occurred. Please reload window and try again. If the problem still persists open an issue at
         
        <a
          class="underline"
          href="https://github.com/bangle-io/bangle-io-issues"
        >
          Github
        </a>
         
      </span>
    </div>
  </div>
</div>
`;

exports[`renders correctly WORKSPACE_AUTH_REJECTED_ERROR 1`] = `
<div>
  <div>
    <div
      class="w-full px-4 text-center"
      data-testid="WORKSPACE_AUTH_REJECTED"
      style="color: white;"
    >
      <span>
        Bangle.io was denied access to your notes. Please try again and press allow Bangle.io access to your locally saved notes.
      </span>
    </div>
  </div>
</div>
`;

exports[`renders correctly WORKSPACE_NAME_ALREADY_EXISTS_ERROR 1`] = `
<div>
  <div>
    <div
      class="w-full px-4 text-center"
      data-testid="WORKSPACE_NAME_ALREADY_EXISTS"
      style="color: white;"
    >
      <span>
        A workspace with the same name already exists.
         
        <button
          class="underline"
        >
          Click here
        </button>
         
        to open it
      </span>
    </div>
  </div>
</div>
`;
